src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -655,78 +655,34 @@ fn resolve_associated_trait_item(
-                        });
-                        let assoc = items.next();
-                        debug_assert_eq!(items.count(), 0);
-                        assoc
-                    } else {
-                        // These are provided methods or default types:
-                        // ```
-                        // trait T {
-                        //   type A = usize;
-                        //   fn has_default() -> A { 0 }
-                        // }
-                        // ```
-                        let trait_ = impl_.trait_.unwrap().def_id().unwrap();
-                        cx.tcx
-                            .associated_items(trait_)
-                            .find_by_name_and_namespace(
-                                cx.tcx,
-                                Ident::with_dummy_span(item_name),
-                                ns,
-                                trait_,
-                            )
-                            .map(|assoc| (assoc.kind, assoc.def_id))
-                    }
-                }
-                _ => panic!("get_impls returned something that wasn't an impl"),
-            }
-        })
-        .collect();
-
+    // FIXME: this should also consider blanket impls (`impl<T> X for T`). Unfortunately
+    // `get_auto_trait_and_blanket_impls` is broken because the caching behavior is wrong. In the
+    // meantime, just don't look for these blanket impls.
+    
     // Next consider explicit impls: `impl MyTrait for MyType`
     // Give precedence to inherent impls.
-    if candidates.is_empty() {
-        let traits = traits_implemented_by(cx, did, module);
-        debug!("considering traits {:?}", traits);
-        candidates.extend(traits.iter().filter_map(|&trait_| {
-            cx.tcx
-                .associated_items(trait_)
-                .find_by_name_and_namespace(cx.tcx, Ident::with_dummy_span(item_name), ns, trait_)
-                .map(|assoc| (assoc.kind, assoc.def_id))
-        }));
-    }
+    let traits = traits_implemented_by(cx, did, module);
+    debug!("considering traits {:?}", traits);
+    let candidates = traits.iter().filter_map(|&trait_| {
+        cx.tcx
+            .associated_items(trait_)
+            .find_by_name_and_namespace(cx.tcx, Ident::with_dummy_span(item_name), ns, trait_)
+            .map(|assoc| (assoc.kind, assoc.def_id))
+    });
     // FIXME(#74563): warn about ambiguity
-    debug!("the candidates were {:?}", candidates);
-    candidates.pop()
+    // NOTE: this looks weird because it only calls `collect()` in debug mode
+    // and `candidates` only needs to be mutable if `next()` called.
+    #[cfg(debug_assertions)]
+    {
+        let mut all_candidates: Vec<_> = candidates.collect();
+        debug!("the candidates were {:?}", all_candidates);
+        all_candidates.pop()
+    }
+    #[cfg(not(debug_assertions))]
+    {
+        let mut candidates = candidates;
+        candidates.next()
+    }
```

Comment:
I can revert this change if you think the performance isn't worth it. I think the logging is valuable since this is actively discarding info, though.

Reply:
Maybe it'd make sense to simpy do:
```rust
debug!("...", candidates.clone().collect::<Vec<_>>());
```

IIUC this should only do expensive operation if `debug` level is enabled (assuming that iterator cloning would behave as I'd expect)

Reply:
ah right, I forgot you can clone iterators. Thanks!